jplag/src/test/java/de/jplag/clustering/ClusteringTest.java

```diff
@@ -0,0 +1,267 @@
+package de.jplag.clustering;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.List;
+import java.util.Scanner;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.apache.commons.math3.linear.Array2DRowRealMatrix;
+import org.apache.commons.math3.linear.RealMatrix;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.mockito.invocation.InvocationOnMock;
+
+import de.jplag.JPlagComparison;
+import de.jplag.Submission;
+import de.jplag.clustering.algorithm.GenericClusteringAlgorithm;
+import de.jplag.clustering.algorithm.SpectralClustering;
+import de.jplag.clustering.preprocessors.CumulativeDistributionFunctionPreprocessor;
+
+public class ClusteringTest {
```

Comment:
This is a comment regarding all tests: When looking only at the coverage of your test cases, it seems that there are a few untested classes: The preprocessor classes, the spectral clustering, the clustering factory, and the metric in the cluster class. Could you please add a few test cases for that?

Reply:
I've added test cases for preprocessor classes, the spectral clustering, the metrics and the calculation of the community strength.
I can't think off a meaningful test for the clustering factory, that is not just entirely replicating what's in there.